contracts/contracts/strategies/balancer/BaseBalancerStrategy.sol

```diff
@@ -0,0 +1,323 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+/**
+ * @title OETH Base Balancer Abstract Strategy
+ * @author Origin Protocol Inc
+ */
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import { IERC20, InitializableAbstractStrategy } from "../../utils/InitializableAbstractStrategy.sol";
+import { IBalancerVault } from "../../interfaces/balancer/IBalancerVault.sol";
+import { IRateProvider } from "../../interfaces/balancer/IRateProvider.sol";
+import { VaultReentrancyLib } from "./VaultReentrancyLib.sol";
+import { IOracle } from "../../interfaces/IOracle.sol";
+import { IVault } from "../../interfaces/IVault.sol";
+import { IWstETH } from "../../interfaces/IWstETH.sol";
+import { IERC4626 } from "../../../lib/openzeppelin/interfaces/IERC4626.sol";
+import { StableMath } from "../../utils/StableMath.sol";
+import "hardhat/console.sol";
+
+abstract contract BaseBalancerStrategy is InitializableAbstractStrategy {
+    using SafeERC20 for IERC20;
+    using StableMath for uint256;
+    IBalancerVault internal immutable balancerVault =
+        IBalancerVault(0xBA12222222228d8Ba445958a75a0704d566BF2C8);
+    address internal immutable stEth =
+        0xae7ab96520DE3A18E5e111B5EaAb095312D7fE84;
+    address internal immutable wstEth =
+        0x7f39C581F595B53c5cb19bD0b3f8dA6c935E2Ca0;
+    address internal immutable frxEth =
+        0x5E8422345238F34275888049021821E8E08CAa1f;
+    address internal immutable sfrxEth =
+        0xac3E018457B222d93114458476f3E3416Abbe38F;
+
+    address internal pTokenAddress;
+    bytes32 internal balancerPoolId;
+    // Full list of all assets as they are present in the Balancer pool
+    address[] internal poolAssetsMapped;
+    // Max withdrawal slippage denominated in 1e18 (1e18 == 100%)
+    uint256 public maxWithdrawalSlippage;
+    // Max deposit slippage denominated in 1e18 (1e18 == 100%)
+    uint256 public maxDepositSlippage;
+    int256[50] private __reserved;
+
+    event MaxWithdrawalSlippageUpdated(
+        uint256 _prevMaxSlippagePercentage,
+        uint256 _newMaxSlippagePercentage
+    );
+    event MaxDepositSlippageUpdated(
+        uint256 _prevMaxSlippagePercentage,
+        uint256 _newMaxSlippagePercentage
+    );
+
+    /**
+     * @dev Ensure we are not in a Vault context when this function is called, by attempting a no-op internal
+     * balance operation. If we are already in a Vault transaction (e.g., a swap, join, or exit), the Vault's
+     * reentrancy protection will cause this function to revert.
+     *
+     * Use this modifier with any function that can cause a state change in a pool and is either public itself,
+     * or called by a public function *outside* a Vault operation (e.g., join, exit, or swap).
+     */
+    modifier whenNotInVaultContext() {
+        VaultReentrancyLib.ensureNotInVaultContext(balancerVault);
+        _;
+    }
+
+    /**
+     * @dev Returns bool indicating whether asset is supported by strategy
```

Comment:
this is an external function so should use @notice so Etherscan will pick up the Natspec.

Reply:
👍